lib/web_ui/lib/src/engine/semantics/semantics.dart

```diff
@@ -1828,7 +1877,23 @@ class EngineSemanticsOwner {
 
     assert(_semanticsTree.containsKey(0)); // must contain root node
     assert(() {
-      // Validate tree
+      // Validate that the node map only contains live elements, i.e. descendants
+      // of the root node. If a node is not reachable from the root, it should
+      // have been removed from the map.
+      final List<int> liveIds = <int>[];
+      final SemanticsObject root = _semanticsTree[0]!;
+      root.visitDepthFirst((SemanticsObject child) {
+        liveIds.add(child.id);
+      });
+      if (!_semanticsTree.keys.every(liveIds.contains)) {
+        throw AssertionError(
+          'The semantics node map is inconsistent:\n'
+          '  Nodes in tree: [${liveIds.join(', ')}]\n'
+          '  Nodes in map : [${_semanticsTree.keys.join(', ')}]'
+        );
+      }
```

Comment:
Because I was under the false impression that assertion error expressions are evaluated eagerly, and this one's really expensive. However, I was proven wrong.

Done.